### PR TITLE
Variable Length Nft Keys

### PR DESCRIPTION
**Pull request by talekhinezh**

### CODE REVIEW

File: scrypto/src/buffer/constants.rs [hunk 1]
```diff
@@ -6,6 +6,7 @@ pub const SCRYPTO_TYPE_BID: u8 = 0x84;
 pub const SCRYPTO_TYPE_RID: u8 = 0x85;
 pub const SCRYPTO_TYPE_MID: u8 = 0x86;
 pub const SCRYPTO_TYPE_VID: u8 = 0x87;
+pub const SCRYPTO_TYPE_NFT_KEY: u8 = 0x88;
```
Comment (iamyulong) [inline scrypto/src/buffer/constants.rs:9]: There are a few more places we need to update to fully support this new type. I'd suggest you do a global search `SCRYPTO_TYPE_VID`  and add corresponding handling of NftKeys.

File: simulator/src/ledger/dumper.rs [hunk 2]
```diff
@@ -119,16 +119,16 @@ fn dump_resources<T: Ledger>(vaults: &HashSet<Vid>, ledger: &T) -> Result<(), Di
...
                 let mutable_data = validate_data(&nft.mutable_data()).unwrap();
                 println!(
                     "{}  {} NFT {{ id: {}, immutable_data: {}, mutable_data: {} }}",
                     if last { " " } else { "│" },
                     list_item_prefix(inner_last),
-                    id,
+                    key,
                     immutable_data,
                     mutable_data
                 );
```
Comment (iamyulong) [inline simulator/src/ledger/dumper.rs:134]: We need to update resim and transaction builder to handle passing of NFT buckets (ideally to support all NFT keys including `u128`).

File: assets/account/src/lib.rs [hunk 3]
```diff
@@ -2,24 +2,24 @@ use scrypto::prelude::*;
 
 blueprint! {
     struct Account {
-        key: Address,
+        address: Address,
```
Comment (iamyulong) [inline assets/account/src/lib.rs:5]: I'd suggest we rename this to `public_key`.

File: scrypto/src/types/nft_key.rs [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+    pub fn new(v: Vec<u8>) -> Self {
+        NftKey(v)
+    }
+}
+
+impl From<u128> for NftKey {
+    fn from(u: u128) -> Self {
+        NftKey(u.to_be_bytes().to_vec())
+    }
+}
```
Comment (iamyulong) [inline scrypto/src/types/nft_key.rs:22]: I'd suggest we add a conversion for any type with `sbor::Encode` trait.

File: scrypto/src/buffer/constants.rs (hunk 1 above)
Comment (iamyulong) [reply scrypto/src/buffer/constants.rs:9]: May also need to update blueprint ABI to handle this type more elegantly.  For now, it's encoded and decoded as `Struct` which is not very convenient.

File: scrypto/src/types/nft_key.rs (hunk 4 above)
Comment (talekhinezh) [reply scrypto/src/types/nft_key.rs:22]: This interface should only support conversions to bytes array of bounded length. `sbor::Encode` would expose too much.

**Review by iamyulong — APPROVED** [review]

**Pull request merged**